### PR TITLE
Indicate that a final newline was added in --diff

### PR DESCRIPTION
**Pull request by TBBle**
Fixes: #1662

This works around [bpo2142](https://bugs.python.org/issue2142) from 2008.

### CODE REVIEW

**Comment by TBBle** [comment]
Now that I know it's an existing Python bug (I didn't think to look until after I posted the PR), I could reword this as a workaround for that bug, if that's valuable.

File: src/black/__init__.py [hunk 1]
```diff
@@ -6380,11 +6380,18 @@ def diff(a: str, b: str, a_name: str, b_name: str) -> str:
...
+    diff_lines = []
+    for line in difflib.unified_diff(
+        a_lines, b_lines, fromfile=a_name, tofile=b_name, n=5
+    ):
+        if line[-1] == "\n":
+            diff_lines.append(line)
+        else:
+            diff_lines.append(line + "\n")
+            diff_lines.append("\\ No newline at end of file\n")
+    return "".join(diff_lines)
```
Comment (ichard26) [inline src/black/__init__.py:6394]: To someone like me, the `diff_lines.append("\\ No newline at end of file\n")` link just looks plain wrong. I would suggest explaining that this code is using a special feature of the patch specification(?) to work around a Python bug. Also linking the BPO would be the cherry on the cake!

**Review by ichard26 — CHANGES_REQUESTED** [review]
The changes themselves look great and testing locally they do work so +1 from me!

Besides my review comment, please add an entry in `CHANGES.md` so the person verifying and updating the changelog before a new release has less work to do :) And I guess while you're at it, why not add yourself to the authors list in the README?

Also while this isn't a blocker, I can't reproduce the bug where Black doesn't add a newline using the VS Code integration on my machine interestingly 😕 Maybe the fact I'm using Visual Studio Code version 1.52.1 or version v2020.12.424452561 of the Python extension is why? I honestly don't care that much since theoretically this should fix the bug, but from experience even seemingly well tested PRs can have undetected serious holes ...

Thanks for looking into this and submitting a patch anyway! You rock!

File: src/black/__init__.py (hunk 1 above)
Comment (TBBle) [reply src/black/__init__.py:6394]: Yeah, sounds good. Now that I have a BPO reference, I can do that.

**Comment by TBBle** [comment]
Would it be useful to mention _CHANGES.md_ in _CONTRIBUTING.md_? I wasn't aware I was supposed to add to the former, based only on reading the latter. (And based on the PR I modelled from, #1328.

**Comment by TBBle** [comment]
@ichard26 Should _CHANGES.md_ link to the PR or the issue?

**Comment by ichard26** [comment]
> Would it be useful to mention CHANGES.md in CONTRIBUTING.md? I wasn't aware I was supposed to add to the former, based only on reading the latter. (And based on the PR I modelled from, #1328.)

Yeah it would definitely help! Although I have other ideas to deal this issue (hi [towncrier](https://pypi.org/project/towncrier/)).

> Should CHANGES.md link to the PR or the issue

Honestly I'm not sure of what the actual recommendation is, but I would go with linking to the issue personally even though the status quo is to link to the PR.

File: src/black/__init__.py (hunk 1 above)
Comment (TBBle) [reply src/black/__init__.py]: Done

File: CHANGES.md [hunk 2]
```diff
@@ -26,6 +26,8 @@
 
 - use lowercase hex strings (#1692)
 
+- `--diff` correctly indicates when a file doens't end in a newline (#1662)
```
Comment (ichard26) [inline CHANGES.md:29]: Not a blocker but:
```suggestion
- `--diff` correctly indicates when a file doesn't end in a newline (#1662)
```

**Review by ichard26 — APPROVED** [review]

**Comment by TBBle** [comment]
@ichard26 On the repro in VS Code, it's possible VS Code's general editor setting for "add missing newline" is set, which will be operating _after_ `black --diff` has been applied. That's the workaround I am using until this fix flows through to a release.

**Comment by TBBle** [comment]
Sorry for the wild batch of force-pushes. I got excited and overlooked a couple of requested changes. -_-

It should be all-good now, the only changes since the last review were [textual](https://github.com/psf/black/compare/05b41c23af52ec4314e08b8553e9abfc7f6eb528..5d148d1cad2502bffd46c1ffda30bdba2fc16da9)

**Comment by ichard26** [comment]
> On the repro in VS Code, it's possible VS Code's general editor setting for "add missing newline" is set

I checked if I had that option set and I'm pretty sure I didn't. When I purposely installed an immediately crashing version of Black, no newline was added still. But it's probably some misconfiguration or misunderstanding on my part, anyway 95% of time I use Vim. Does this patch for you work on your Visual Studio Code setup? If yes, I'll just go with that then :-)

**Comment by TBBle** [comment]
I can reproduce the problem with released Black and Python extension ms-python.python-2020.12.424452561, under VSCode 1.52.1, after I remove `"files.insertFinalNewline": true` from the `[python]` section of settings.json (which is not visible in the settings UI editor).

And using the black.exe built from this PR, the problem is fixed.

**Comment by ichard26** [comment]
Looking at this again, somethings are sticking out to me, I'mma do another review hopefully soon.

File: src/black/__init__.py [hunk 3]
```diff
@@ -6424,11 +6424,20 @@ def diff(a: str, b: str, a_name: str, b_name: str) -> str:
...
+        a_lines, b_lines, fromfile=a_name, tofile=b_name, n=5
+    ):
+        # Work around https://bugs.python.org/issue2142
+        # See https://www.gnu.org/software/diffutils/manual/html_node/Incomplete-Lines.html
+        if line[-1] == "\n":
+            diff_lines.append(line)
+        else:
+            diff_lines.append(line + "\n")
+            diff_lines.append("\\ No newline at end of file\n")
+    return "".join(diff_lines)
```
Comment (ichard26) [inline src/black/__init__.py:6440]: Taking a second look at this, this is confusing IMO. `\\ No newline at end of file\n` means to me that the else block should only happen at the end of the diff, but the code is written like this condition can happen wherever in the diff. I would appreciate if this could be clarified. If this extra code is only useful for diff's ending, would be it possible to just check and modify the diff's ending like this (just illustrative code):
```py3
diff_lines = unified_diff(a, b, ...)
last_line = diff_lines[-1]
if last_line[-1] != "\n":
    diff_lines[-1] = last_line + "\n"
    diff_lines.append("\\ No newline at end of file\n")
```

If it's the other way around, could the `\\ No newline at end of file\n` message be tweaked?

**Review by ichard26 — COMMENTED** [review]
Hey this almost looks good to go, but on a second pass, something doesn't seem right. Not a big deal, just something odd I noticed.

File: src/black/__init__.py (hunk 3 above)
Comment (TBBle) [reply src/black/__init__.py:6440]: Per [the spec](https://www.gnu.org/software/diffutils/manual/html_node/Incomplete-Lines.html), only the last line of a file can be the _incomplete line_, yes.

However, if you pass in non-final lines without newlines, `difflib` will accept that, and happily produce output lines without newlines.

Since that latter effect is really Garbage-In-Garbage-Out, and we're using `keepends` now to ensure that cannot happen, your suggested change makes sense, yes.

Comment (TBBle) [reply src/black/__init__.py:6464]: Actually, wait. `diff_lines` doesn't necessarily have the last line of the input file as its last line. In our case, since we always add a final newline, it's only the last `-...` line that may hit this case, not the last line in `diff_lines`, which will be the corresponding `+...` line.

See tests/data/missing_final_newline.diff for this exact case.

In the general case, a _context_ line (` ...`) or added line (`+ ...`) could require this marker too.

Comment (TBBle) [reply src/black/__init__.py:6464]: Technically, I could walk backwards in `diff_lines`, checking the final `-`, final `+`, and final ` ` for missing final newline (and if it's the ` ` case, barfing if we also see it on a `+` or `-` case) but if I was going to do that, I should also be checking for a `\` line, which would appear if BPO 2142 is fixed, and hence this workaround would be deactivated in that case.

Comment (TBBle) [reply src/black/__init__.py:6464]: If the text of the `\...` line is confusing, that can be trivially changed, this is just copied from the spec, but in the actual implementation of `diff`, it's localised with gettext. (Although I fear that there are patch-parsers in the wild which are looking for _this text_ rather than the `\` marker.)

We could even look at the first character of the affected line, and have the text vary by whether it was a `-...`, `+...`, or ` ...` line.

I won't do that without some positive direction though, as it feels like it's adding needless complexity for what is a pretty-standard text string, i.e. `git diff` also produces this literal string. Even GitHub's diff preview uses the exact string "No newline at end of file" in its popup. And it's complexity that a future fix for BPO 2142 probably won't have, since the existing proposed fixes didn't try to distinguish this either. So the patch format would revert back to what it is here, at that point.

Comment (ichard26) [reply src/black/__init__.py:6464]: Alright that makes sense. Thanks for describing the specification!

**Review by ichard26 — APPROVED** [review]
/cc @cooperlees @JelleZijlstra this should be ready to go, a final lookover and eventual merge would be appreciated! Thanks in advance!

**Comment by TBBle** [comment]
Master build failed on [Documentation build](https://github.com/psf/black/runs/1949719732) but I can't see how it relates to my changes.

```
Warning, treated as error:
autodoc: failed to import function 'should_split_body_explode' from module 'black'; the following exception was raised:
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sphinx/util/inspect.py", line 325, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'black' has no attribute 'should_split_body_explode'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sphinx/ext/autodoc/importer.py", line 106, in import_object
    obj = attrgetter(obj, mangled_name)
  File "/opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sphinx/ext/autodoc/__init__.py", line 289, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sphinx/ext/autodoc/__init__.py", line 2186, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sphinx/util/inspect.py", line 341, in safe_getattr
    raise AttributeError(name) from exc
AttributeError: should_split_body_explode
```

`should_split_body_explode` was renamed to `should_split_line` in #1963, and I guess _/docs/reference/reference_functions.rst_ wasn't updated.

**Comment by jayaddison** [comment]
@TBBle Out of interest, did you spot the newline handling at https://github.com/psf/black/blob/cd4295dd9888f491cfd9aae060d7832b7f831b24/src/black/__init__.py#L1116-L1117 while preparing your changes here?  I don't think they're directly related, but wondered if we might be able to remove that special case at some point.

**Comment by TBBle** [comment]
That code looks more like it's related to the parser driver requiring newline-terminated text, which I think is probably part of the API contract.
